File: packages/nativescript-image-cache-it/src-native/imagecacheit/src/main/java/com/github/triniwiz/imagecacheit/AspectTransformer.java
package com.github.triniwiz.imagecacheit;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Shader;

import androidx.annotation.NonNull;

import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.request.target.Target;

import java.security.MessageDigest;

import jp.wasabeef.glide.transformations.BitmapTransformation;

/**
 * Created by triniwiz on 2019-06-28
 */
public class AspectTransformer extends BitmapTransformation {
    private static final int VERSION = 1;
    private static final String ID = "com.github.triniwiz.imagecacheit.AspectTransformer." + VERSION;

    private int decodeWidth;
    private int decodeHeight;
    private boolean keepAspectRatio;

    public AspectTransformer(int decodeWidth, int decodeHeight, boolean keepAspectRatio) {
        this.decodeHeight = decodeHeight;
        this.decodeWidth = decodeWidth;
        this.keepAspectRatio = keepAspectRatio;
    }

    @Override
    protected Bitmap transform(@NonNull Context context, @NonNull BitmapPool pool,
                               @NonNull Bitmap toTransform, int outWidth, int outHeight) {


        Bitmap bitmap = toTransform;
        int sourceWidth = toTransform.getWidth();
        int sourceHeight = toTransform.getHeight();
        int reqWidth = decodeWidth > 0 ? decodeWidth : Target.SIZE_ORIGINAL;
        int reqHeight = decodeHeight > 0 ? decodeHeight : Target.SIZE_ORIGINAL;

        // scale
        if ((reqWidth != Target.SIZE_ORIGINAL && reqHeight != Target.SIZE_ORIGINAL) && (reqWidth != sourceWidth || reqHeight != sourceHeight)) {
            if (keepAspectRatio) {
                double widthCoef = (double) sourceWidth / (double) reqWidth;
                double heightCoef = (double) sourceHeight / (double) reqHeight;
                double aspectCoef = Math.min(widthCoef, heightCoef);

                reqWidth = (int) Math.floor(sourceWidth / aspectCoef);
                reqHeight = (int) Math.floor(sourceHeight / aspectCoef);
            }

            bitmap = Bitmap.createScaledBitmap(bitmap, reqWidth, reqHeight, true);
        }

        return bitmap;
    }

    @Override
    public String toString() {
        return "AspectTransformer(decodeWidth=" + decodeWidth + ", decodeHeight=" + decodeHeight + ", keepAspectRatio="
                + keepAspectRatio + ")";
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof AspectTransformer &&
                ((AspectTransformer) o).decodeWidth == decodeWidth &&
                ((AspectTransformer) o).decodeHeight == decodeHeight &&
                ((AspectTransformer) o).keepAspectRatio == keepAspectRatio;
    }

    @Override
    public int hashCode() {
        return ID.hashCode();
    }

    @Override
    public void updateDiskCacheKey(@NonNull MessageDigest messageDigest) {
        messageDigest.update((ID + decodeWidth + decodeHeight).getBytes(CHARSET));
    }
}

File: packages/nativescript-image-cache-it/src-native/android/imagecacheit/src/main/java/com/github/triniwiz/imagecacheit/AspectTransformer.kt
package com.github.triniwiz.imagecacheit

import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.request.target.Target
import jp.wasabeef.glide.transformations.BitmapTransformation
import java.security.MessageDigest
import kotlin.math.floor

/**
 * Created by triniwiz on 2019-06-28
 */
class AspectTransformer(private val decodeWidth: Int, private val decodeHeight: Int, private val keepAspectRatio: Boolean) : BitmapTransformation() {
  override fun transform(context: Context, pool: BitmapPool,
                         toTransform: Bitmap, outWidth: Int, outHeight: Int): Bitmap {
    var bitmap = toTransform
    val sourceWidth = toTransform.width
    val sourceHeight = toTransform.height
    var reqWidth = if (decodeWidth > 0) decodeWidth else Target.SIZE_ORIGINAL
    var reqHeight = if (decodeHeight > 0) decodeHeight else Target.SIZE_ORIGINAL

    // scale
    if (reqWidth != Target.SIZE_ORIGINAL && reqHeight != Target.SIZE_ORIGINAL && (reqWidth != sourceWidth || reqHeight != sourceHeight)) {
      if (keepAspectRatio) {
        val widthCoef = sourceWidth.toDouble() / reqWidth.toDouble()
        val heightCoef = sourceHeight.toDouble() / reqHeight.toDouble()
        val aspectCoef = widthCoef.coerceAtMost(heightCoef)
        reqWidth = floor(sourceWidth / aspectCoef).toInt()
        reqHeight = floor(sourceHeight / aspectCoef).toInt()
      }
      bitmap = Bitmap.createScaledBitmap(bitmap, reqWidth, reqHeight, true)
    }
    return bitmap
  }

  override fun toString(): String {
    return ("AspectTransformer(decodeWidth=" + decodeWidth + ", decodeHeight=" + decodeHeight + ", keepAspectRatio="
      + keepAspectRatio + ")")
  }

  override fun equals(o: Any?): Boolean {
    return o is AspectTransformer && o.decodeWidth == decodeWidth && o.decodeHeight == decodeHeight && o.keepAspectRatio == keepAspectRatio
  }

  override fun hashCode(): Int {
    return ID.hashCode()
  }

  override fun updateDiskCacheKey(messageDigest: MessageDigest) {
    messageDigest.update((ID + decodeWidth + decodeHeight).toByteArray(CHARSET))
  }

  companion object {
    private const val VERSION = 1
    private const val ID = "com.github.triniwiz.imagecacheit.AspectTransformer.$VERSION"
  }
}
